File: tensorflow/tensorboard/components/tf-graph-common/test/util-test.ts
/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the 'License');
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an 'AS IS' BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

suite('util', () => {
  let assert = chai.assert;

  test('remove common prefix', () => {

    // Empty array.
    let result = tf.graph.util.removeCommonPrefix([]);
    assert.deepEqual(result, []);

    // No common prefix.
    result = tf.graph.util.removeCommonPrefix(['a', 'b', 'c']);
    assert.deepEqual(result, ['a', 'b', 'c']);

    // One of the elements is empty string.
    result = tf.graph.util.removeCommonPrefix(['a/b', '', 'a/c']);
    assert.deepEqual(result, ['a/b', '', 'a/c']);

    // Only one string.
    result = tf.graph.util.removeCommonPrefix(['a/b/c']);
    assert.deepEqual(result, ['a/b/c']);

    // `q/w/` is the common prefix. Expect `q/w/` to be removed.
    result = tf.graph.util.removeCommonPrefix(['q/w/a', 'q/w/b', 'q/w/c/f']);
    assert.deepEqual(result, ['a', 'b', 'c/f']);

    // `q/w/` is the common prefix and also an element. Expect nothing to be
    // removed since the common prefix is also an element in the array.
    result = tf.graph.util.removeCommonPrefix(['q/w/', 'q/w/b', 'q/w/c/f']);
    assert.deepEqual(result, ['q/w/', 'q/w/b', 'q/w/c/f']);
  });

  test('query params', () => {
    // Starts with question mark.
    let queryParams = tf.graph.util.getQueryParams('?foo=1&bar=2');
    assert.deepEqual(queryParams, {'foo': '1', 'bar': '2'});

    // No question mark.
    queryParams = tf.graph.util.getQueryParams('foo=1&bar=2');
    assert.deepEqual(queryParams, {'foo': '1', 'bar': '2'});
  });
});

File: tensorflow/lite/experimental/delegates/coreml/builders/util_test.cc
/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/
#include "tensorflow/lite/experimental/delegates/coreml/builders/util.h"

#include <algorithm>

#include <gmock/gmock.h>
#include <gtest/gtest.h>
#include "tensorflow/lite/c/common.h"

using tflite::delegates::coreml::IsBinaryOpSupported;

namespace {

class IsBinaryOpSupportedTest : public testing::Test {
 protected:
  void SetUp() override {
    const int input_size = 2;
    tensors_.resize(input_size);
    context_.tensors = tensors_.data();
    node_.inputs = TfLiteIntArrayCreate(input_size);
    for (int i = 0; i < input_size; ++i) {
      node_.inputs->data[i] = i;
    }

    for (auto& tensor : tensors_) {
      tensor.allocation_type = kTfLiteArenaRw;
      tensor.dims = nullptr;
    }
  }

  void TearDown() override {
    FreeInputShapes();
    TfLiteIntArrayFree(node_.inputs);
  }

  void SetInputShapes(const std::vector<std::vector<int>>& shapes) {
    for (int i = 0; i < tensors_.size(); ++i) {
      tensors_[i].dims = TfLiteIntArrayCreate(shapes[i].size());
      std::copy(shapes[i].begin(), shapes[i].end(), tensors_[i].dims->data);
    }
  }

  void FreeInputShapes() {
    for (auto& tensor : tensors_) {
      if (tensor.dims != nullptr) {
        TfLiteIntArrayFree(tensor.dims);
        tensor.dims = nullptr;
      }
    }
  }

  TfLiteContext context_;
  TfLiteNode node_;
  std::vector<TfLiteTensor> tensors_;
};

TEST_F(IsBinaryOpSupportedTest, BroadcastTest) {
  std::vector<int> base_shape = {2, 2, 3, 4};
  std::vector<std::vector<int>> shapes = {
      {2, 2, 3, 4}, {2, 1, 1, 4}, {2, 2, 3, 1}, {2, 1, 1, 1}};
  std::vector<TfLiteTensor> inputs(2);
  for (const auto& shape : shapes) {
    SetInputShapes({base_shape, shape});
    ASSERT_TRUE(IsBinaryOpSupported(nullptr, &node_, &context_));
    FreeInputShapes();
  }
}

TEST_F(IsBinaryOpSupportedTest, LessThan4DTest) {
  std::vector<int> base_shape = {1, 2, 3, 4};
  std::vector<std::vector<int>> shapes = {{4}, {2, 3, 1}, {1, 1, 1, 1}};
  for (const auto& shape : shapes) {
    SetInputShapes({base_shape, shape});
    ASSERT_TRUE(IsBinaryOpSupported(nullptr, &node_, &context_));
    FreeInputShapes();
  }
}

TEST_F(IsBinaryOpSupportedTest, ConstScalarTest) {
  std::vector<int> base_shape = {2, 2, 3, 4};
  tensors_[1].allocation_type = kTfLiteMmapRo;
  SetInputShapes({base_shape, {1}});
  ASSERT_TRUE(IsBinaryOpSupported(nullptr, &node_, &context_));
  FreeInputShapes();
}

TEST_F(IsBinaryOpSupportedTest, NotSupportedBroadcastTest) {
  std::vector<int> base_shape = {2, 2, 3, 4};
  std::vector<std::vector<int>> shapes = {
      {2, 2, 1, 4}, {2, 1, 2, 4}, {1, 2, 3, 1}, {1, 1, 1, 1}};
  for (const auto& shape : shapes) {
    SetInputShapes({base_shape, shape});
    ASSERT_FALSE(IsBinaryOpSupported(nullptr, &node_, &context_));
    FreeInputShapes();
  }
}
}  // namespace
